opendevin/controller/agent_controller.py

```diff
@@ -196,25 +217,46 @@ async def _step(self):
             await asyncio.sleep(1)
             return
 
-        logger.info(f'STEP {self.state.iteration}', extra={'msg_type': 'STEP'})
-        if self.state.iteration >= self.max_iterations:
-            await self.report_error('Agent reached maximum number of iterations')
-            await self.set_agent_state_to(AgentState.ERROR)
-            return
-
         if self.delegate is not None:
-            delegate_done = await self.delegate._step()
+            await self.delegate._step()
+            delegate_state = self.delegate.get_agent_state()
+            if delegate_state == AgentState.ERROR:
+                # close the delegate upon error
+                await self.delegate.close()
+                await self.report_error('Delegator agent encounters an error')
+                # propagate error state until an agent or user can handle it
+                await self.set_agent_state_to(AgentState.ERROR)
+                return
+            delegate_done = delegate_state == AgentState.FINISHED
             if delegate_done:
+                # retrieve delegate result
                 outputs = self.delegate.state.outputs if self.delegate.state else {}
+
+                # update current controller's state
+                self.state.iteration = self.delegate.state.iteration
+                self.state.num_of_chars = self.delegate.state.num_of_chars
+
+                # close delegate controller: we must close the delegate controller before adding new events
+                await self.delegate.close()
+
+                # update delegate result observation
                 obs: Observation = AgentDelegateObservation(content='', outputs=outputs)
                 await self.event_stream.add_event(obs, EventSource.AGENT)
+
+                # clean up delegate status
                 self.delegate = None
                 self.delegateAction = None
             return
 
         if self.state.num_of_chars > self.max_chars:
             raise MaxCharsExceedError(self.state.num_of_chars, self.max_chars)
 
+        logger.info(f'STEP {self.state.iteration}', extra={'msg_type': 'STEP'})
+        if self.state.iteration >= self.state.max_iterations:
+            await self.report_error('Agent reached maximum number of iterations')
+            await self.set_agent_state_to(AgentState.ERROR)
+            return
```

Comment:
I think this should be `STOPPED` rather than `ERROR`--the user may very well want to keep going

Reply:
We have an issue for this: https://github.com/OpenDevin/OpenDevin/issues/1493